docs: clarify dynamic imports and side effects (#31479)
nodejs/node: doc/api/esm.md

import { method } from 'commonjs-package'; // Errors
```

It is also possible to import the top-level code (side effects) only:

```js
import 'package-or-file';
```

### `import()` expressions

Dynamic `import()` is supported in both CommonJS and ES modules. It is useful

> when you need to do one of the following:

Reviewer: ```suggestion
Dynamic `import()` is supported in both CommonJS and ES modules. It is useful
when needing to do one of the following:
```

* import from an ES module file in CommonJS code
* import an ES module file conditionally in either CommonJS or ES module code

Dynamic import is useful when the module has side effects that you want only
if some logic is true. By "side effects", we mean top-level code that would
execute just by importing the file, even if you never call the function you
import. You should avoid writing ES modules with side effects, but they may

> exist in a package you depend on, and it's useful to know how to avoid them.

Reviewer: ```suggestion
Dynamic import is useful when the module has side effects only if some logic
is true. Here "side effects" means top-level code that would execute just by
importing the file, even if the imported function is never called. ES modules
should not be implemented with such side effects, but they may exist in an
dependency.
```

In general, avoid the use of `you`, `we`, etc in the api docs. It's entirely possible that we'll revisit that rule later but that's the style guideline

Beyond just avoiding side effects, importing dynamically may also decrease a
program's memory consumption. However, it makes static dependency analysis

> and tree shaking more difficult, so you need to weigh the pros and cons.

Reviewer: ```suggestion
and tree shaking more difficult.
```

import { method } from 'commonjs-package'; // Errors
```

It is also possible to
[import an ES or CommonJS module for its side effects only][].

### `import()` expressions

[Dynamic import][] is supported in both CommonJS and ES modules. It is the

> _only_  way to import from an ES module in CommonJS code.

Reviewer 2: ```suggestion
[Dynamic `import()`][] is supported in both CommonJS and ES modules. It can be used
to include ES module files from CommonJS code.
```

Technically it’s not the only way; there are `data:` URIs, `Worker`s (I think) and possibly other methods.

[Babel]: https://babeljs.io/
[CommonJS]: modules.html
[Conditional Exports]: #esm_conditional_exports

> [Dynamic import]: https://wiki.developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#Dynamic_Imports

Reviewer 2: ```suggestion
[Dynamic `import()`]: https://wiki.developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#Dynamic_Imports
```
